pkg/inject/inject.go

```diff
@@ -814,35 +821,58 @@ func (conf *ResourceConfig) proxyOutboundSkipPorts() string {
-// workload should be injected. It shouldn't if:
-// - it contains any known sidecars; or
-// - is on a HostNetwork; or
-// - the pod is annotated with "linkerd.io/inject: disabled".
-func ShouldInjectCLI(_ *ResourceConfig, r Report) bool {
-	return r.Injectable()
+// HasPayload returns true if the pod's meta isn't empty.
+func (conf *ResourceConfig) HasPayload() bool {
+	return conf.pod.meta.ObjectMeta != nil && conf.pod.spec != nil
 }
 
-// ShouldInjectWebhook determines whether or not the given workload should be
-// injected. It shouldn't if:
-// - it contains any known sidecars; or
-// - is on a HostNetwork; or
-// - the pod is annotated with "linkerd.io/inject: disabled".
-// Additionally, a workload should be injected if:
-// - the workload's namespace has the linkerd.io/inject annotation set to
-//   "enabled", and the workload's pod spec does not have the
-//   linkerd.io/inject annotation set to "disabled"; or
-// - the workload's pod spec has the linkerd.io/inject annotation set to "enabled"
-func ShouldInjectWebhook(conf *ResourceConfig, r Report) bool {
-	if !r.Injectable() {
-		return false
-	}
+// InjectEnabled returns true if the linkerd.io/inject=enabled annotation is
+// present either at the namespace or pod level, without a counter
+// linkerd.io/inject=disabled annotation.
+func (conf *ResourceConfig) InjectEnabled() bool {
+	podAnnotation := conf.pod.meta.Annotations[k8s.ProxyInjectAnnotation]
 
-	podAnnotation := conf.podMeta.Annotations[k8s.ProxyInjectAnnotation]
-	nsAnnotation := conf.nsAnnotations[k8s.ProxyInjectAnnotation]
-	if nsAnnotation == k8s.ProxyInjectEnabled && podAnnotation != k8s.ProxyInjectDisabled {
-		return true
+	if conf.nsAnnotations != nil {
+		nsAnnotation := conf.nsAnnotations[k8s.ProxyInjectAnnotation]
+		if nsAnnotation == k8s.ProxyInjectEnabled && podAnnotation != k8s.ProxyInjectDisabled {
+			return true
+		}
 	}
 
 	return podAnnotation == k8s.ProxyInjectEnabled
 }
+
+// InjectDisabled returns true if the workload has a 'linkerd.io/inject:
+// disabled' annotation. This is one of the conditions used by the CLI to
+// determine if a proxy injection should happen.
+func (conf *ResourceConfig) InjectDisabled() bool {
+	return conf.pod.meta.ObjectMeta.GetAnnotations()[k8s.ProxyInjectAnnotation] == k8s.ProxyInjectDisabled
+}
+
+// PodUsingHostNetwork returns true if the HostNetwork property is true.
+func (conf *ResourceConfig) PodUsingHostNetwork() bool {
+	return conf.pod.spec != nil && conf.pod.spec.HostNetwork
+}
+
+// HasExistingProxy returns true if the pod already has a proxy sidecar.
+func (conf *ResourceConfig) HasExistingProxy() bool {
+	return conf.pod.spec != nil && healthcheck.HasExistingSidecars(conf.pod.spec)
+}
+
+// ShouldOverrideConfig returns true if the workload has existing sidecars and
+// config override annotations.
+func (conf *ResourceConfig) ShouldOverrideConfig() bool {
+	return healthcheck.HasExistingSidecars(conf.pod.spec) && hasOverrideAnnotations(conf.pod.meta)
```

Comment:
Rather than only doing the override if at least `config.linkerd.io` annotation is present, I think it would better to use something like `k8s.IsMeshed` to determine if the pod is inject, and then always override if inject, even if it doesn't contain override annotations. My thinking is that eventually we might allow users to change the contents of the proxy configmap after install, and we won't pick up the change if we're only looking for override annotations.